### PR TITLE
Initial SFT

### PR DESCRIPTION
**Pull request by ebsmothers**
A merge of various tune components + titan components + stuff from https://github.com/pytorch/torchtitan/pull/1387 + some local hacks to titan (https://github.com/ebsmothers/torchtitan/tree/forge).

Tons of feature gaps, tons of not-thought-out design decisions, but it does run.

```
forge run --nproc_per_node 2 apps/sft/sft.py --config apps/sft/configs/llama3_8b.yaml
```

Enumerating some of the biggest gaps (roughly stack-ranked). Will try to keep this updated, but I claim pretty much all of these can be done in follow-ups.
- [ ] No checkpoint saving
- [ ] No gradient accumulation
- [ ] Currently using causal mask for attention instead of block causal
- [ ] Currently requires packed seq len == model max seq len (discussed offline and we will keep this assumption for now)
- [ ] No normalization of loss by # of unmasked tokens
- [ ] All data setup is currently hardcoded
- [ ] No metric logging
- [ ] Hack to make config system work with titan's literal fields
- [ ] No validation support

### CODE REVIEW

File: apps/sft/sft.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        )
+        dataset = alpaca_iterable_dataset(
+            model_transform=tokenizer,
+            path="yahma/alpaca-cleaned",
+        )
+        min_seq_len_divisor = 2 * self.parallel_dims.tp * self.parallel_dims.cp
+        collate_fn = partial(padded_collate_sft, pad_to_multiple_of=min_seq_len_divisor)
+        # TODO: check this
+        dp_dim = self.parallel_dims.dp_replicate * self.parallel_dims.dp_shard
+        dataloader = StatefulDataLoader(
```
Comment (Darktex) [inline apps/sft/sft.py:107]: You should import Felipe's code for this so we do packing from the get-go

File: apps/sft/sft.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+            total=self.num_training_steps,
+            desc=f"{self.current_step}",
+        )
+
+        while self.current_step < self.num_training_steps:
+            batch = next(dataloader)
+            # Move tensors to the appropriate device
+            for k, v in batch.items():
+                if isinstance(v, torch.Tensor):
+                    batch[k] = v.to("cuda")  # TODO: hardcoded for now
```
Comment (Darktex) [inline apps/sft/sft.py:207]: How does multinode work?

File: forge/data/collate.py [hunk 3]
```diff
@@ -6,6 +6,87 @@
 from typing import Any, Callable
 
 import torch
+import torch.nn.functional as F
+from torch.nn.utils.rnn import pad_sequence
+
+CROSS_ENTROPY_IGNORE_IDX = -100
+
+
+def padded_collate_sft(
```
Comment (Darktex) [inline forge/data/collate.py]: This we should completely remove. No padding, we should support only packing -- and we already have it!

File: forge/data/transforms/utils.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+]
+
+
+class MaskingStrategy(Enum):
+    TRAIN_ON_ALL = "train_on_all"
+    TRAIN_ON_ASSISTANT = "train_on_assistant"
+    TRAIN_ON_LAST = "train_on_last"
+
+
+class Message:
```
Comment (Darktex) [inline forge/data/transforms/utils.py:25]: Weren't we taking the Huggingface messages over?

Comment (Darktex) [reply forge/data/transforms/utils.py:25]: This is the Tune messages right?

File: forge/data/utils.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+Role = Literal[
+    "system",  # Origin is system prompt
+    "user",  # Origin is user
+    "assistant",  # Origin is the model output
+    "ipython",  # Origin is return from a tool call
+    "tool",  # Origin is return from a tool call
+]
+
+
+def padded_collate_sft(
```
Comment (Darktex) [inline forge/data/utils.py]: There are two copypastas of this same function. The other one is in data/collate.py

And we should nuke both! :D

File: forge/datasets/alpaca.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+from typing import Any, Callable, Optional
+
+from forge.data.transforms.tokenizers import ModelTokenizer
+from forge.data.transforms.transforms import AlpacaToMessages
+from forge.datasets.hf import HfIterableDataset
+from forge.datasets.sft import sft_iterable_dataset
+
+
+def alpaca_iterable_dataset(
```
Comment (Darktex) [inline forge/datasets/alpaca.py]: Works but use Felipe's class instead

File: apps/sft/sft.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        model_parts = self.model_parts
+        parallel_dims = self.parallel_dims
+
+        # apply context parallelism if cp is enabled
+        # ensure CP handles the separate freqs_cis buffer for each pp stage
+        inputs = input_dict["tokens"]
+        optional_context_parallel_ctx = (
+            dist_utils.create_context_parallel_ctx(
+                cp_mesh=parallel_dims.world_mesh["cp"],
+                cp_buffers=[inputs, labels] + [m.freqs_cis for m in model_parts],
```
Comment (Darktex) [inline apps/sft/sft.py:133]: Can you do these calculations inside the function and just send down model_parts?

File: apps/sft/sft.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+                    self.pp_schedule.step(
+                        inputs, target=targets, losses=losses, input_batch=inputs
+                    )
+                else:
+                    self.pp_schedule.step(
+                        target=targets, losses=losses, input_batch=inputs
+                    )
+
+            # accumulate losses across pipeline microbatches
+            # TODO: PP+FSDP unexpectedly puts the loss back to the CPU
```
Comment (Darktex) [inline apps/sft/sft.py:158]: Does this also happen on Titan? This code should be super close to a copypasta of what Titan is doing

File: apps/sft/sft.py (hunk 1 above)
Comment (ebsmothers) [reply apps/sft/sft.py:107]: Yeah this is one of the first things I want to do. We will need to integrate this with titan components (basically hotswap flex and do a check that CP is not enabled)

File: apps/sft/sft.py (hunk 2 above)
Comment (ebsmothers) [reply apps/sft/sft.py:207]: Doesn't yet. I just ran this on my dev machine, nothing on MAST yet

File: forge/data/collate.py (hunk 3 above)
Comment (ebsmothers) [reply forge/data/collate.py]: Yeah agreed, responded to your other comment about this above

File: forge/data/transforms/utils.py (hunk 4 above)
Comment (ebsmothers) [reply forge/data/transforms/utils.py:25]: Yeah this is just to unblock. I think we can redesign a bit here

File: forge/datasets/alpaca.py (hunk 6 above)
Comment (ebsmothers) [reply forge/datasets/alpaca.py]: You mean [this one](https://github.com/pytorch-labs/forge/blob/4ea5b2c657bdb1d6793f59633d7a9403643d8121/forge/datasets/sft.py#L15), right? This is basically just a wrapper, but I am OK to take it out if you think it's not providing much value

File: apps/sft/sft.py (hunk 8 above)
Comment (ebsmothers) [reply apps/sft/sft.py:158]: Yes this is a copypasta for now. I think we need to handle loss normalization + padding tokens a bit differently though.

File: apps/sft/sft.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        return loss
+
+    def train_step(self, batch) -> None:
+        # TODO
+        # with GradientAccumulation(
+        #     self.gradient_accumulation_steps,
+        #     self.model,
+        #     self.data_parallel_size,
+        # ) as grad_acc:
+        labels = batch.pop("labels")
```
Comment (felipemello1) [inline apps/sft/sft.py:184]: batch may have other keys nows like metrics and masks to unpack sentences, eg doc_id. Need to make sure that 'forward_backward' is ok with it. In tune, i just added '**kwargs' to the 'forward' call, so those were no opp.

File: apps/sft/sft.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        self.optimizers.zero_grad()
+
+        self.pbar = tqdm(
+            initial=0,
+            total=self.num_training_steps,
+            desc=f"{self.current_step}",
+        )
+
+        while self.current_step < self.num_training_steps:
+            batch = next(dataloader)
```
Comment (felipemello1) [inline apps/sft/sft.py:203]: batch will have 'metrics'. Prob need to add logger here too. There are references in my tune PR.

File: apps/sft/sft.py
Comment (felipemello1) [inline apps/sft/sft.py:1]: pretty nice that stf is only 230 lines. Hopefully its still intuitive to the user how to hack it if they want to.

File: forge/_cli/download.py
Comment (felipemello1) [inline forge/_cli/download.py:1]: might want to delete /scripts/download.py, if this is here

File: apps/sft/sft.py
Comment (ebsmothers) [reply apps/sft/sft.py:1]: Yeah the one thing is that many instance variables are defined inside of titan, so the user will need to know and understand those

File: apps/sft/sft.py (hunk 10 above)
Comment (ebsmothers) [reply apps/sft/sft.py:203]: Yeah haven't touched logging yet, but that can come soon after this initial version

File: forge/datasets/alpaca.py (hunk 6 above)
Comment (Darktex) [reply forge/datasets/alpaca.py]: Yeah I'd just use that one for now, but it's a minor thing either way

File: apps/sft/sft.py (hunk 7 above)
Comment (ebsmothers) [reply apps/sft/sft.py:133]: You mean the fact that we have to explicitly pass the buffers like this? If so I agree, this is just what the [titan API](https://github.com/pytorch/torchtitan/blob/d69a7378cc70e93b1061c9d68ff1c8d54e1594d9/torchtitan/distributed/utils.py#L155-L161) requires. Personally I modified it a bit when reimplementing in tune, is [this](https://github.com/pytorch/torchtune/blob/a036e66842c18d08ee3638c55ed4bc45e944cb4d/torchtune/training/_distributed.py#L807-L812) more what you have in mind?

Comment (ebsmothers) [reply apps/sft/sft.py:133]: More importantly there is also this bug in tune related to CP: https://github.com/pytorch/torchtune/pull/2879. Basically if the seq len in the current batch does not match the model's max seq len (which will happen if we're not packing, and even if we are packing depending on the exact strategy), if we naively shard the RoPE buffers identically to the model inputs we will wind up with a misalignment of RoPE positions. The tune user fixed by leaving the RoPE buffers unsharded and passing input_pos to the model forward to do the correct indexing. However, input_pos is not an arg in titan's transformer implementation, so this won't work here. The way I see it, there are a few paths to handle this:

1) Go all in on packed, always require model seq len == batch seq len
2) Try to extend titan's model implementation to have an equivalent fix to the tune PR
3) Maybe some changes to CP APIs to support custom user-specified sharding of buffers?

(1) somewhat limits the generalizability of some of the packing APIs that @felipemello1 had worked on. It also further raises the visibility of the CP + flex issue -- without support for naively padded batches we likely wind up forcing people to use the technically incorrect pure causal mask until flex CP support lands. (2) sounds like a pain for titan, and it's not yet obvious to me that (3) is even feasible/advisable. Tagging a few folks for thoughts: @Darktex @tianyu-l @XilunWu @pbontrager (who also has an old dynamic batching proposal from last year as an alternative to packing that we could resuscitate, but that's a whole other can of worms).

File: forge/datasets/alpaca.py (hunk 6 above)
Comment (ebsmothers) [reply forge/datasets/alpaca.py]: OK deleted this so marking as resolved

File: forge/data/transforms/utils.py (hunk 4 above)
Comment (ebsmothers) [reply forge/data/transforms/utils.py:25]: @Darktex do you think this is a blocker for this PR? Personally I think there are enough headaches with titan interface, getting the training loop right, etc that I'd personally punt this refactor for a follow-up (since things work with this, that's kinda my top priority atm). Lmk if that sounds reasonable to you though

File: forge/datasets/packed.py [hunk 11]
```diff
@@ -630,7 +637,7 @@ def finalize_pack(
                 if pack["input_pos"]
                 else torch.empty(0, dtype=torch.long)
             ),
-            "metrics": pack["metrics"],
+            # "metrics": pack["metrics"],
```
Comment (felipemello1) [inline forge/datasets/packed.py:640]: why comment it out? i think its fine to keep it. But in the recipe you have to pop them. IMO a better solution.

File: apps/sft/sft.py [hunk 12]
```diff
@@ -85,18 +86,24 @@ def setup_data(self):
...
             model_transform=tokenizer,
+            message_transform=AlpacaToMessages(),
             path="yahma/alpaca-cleaned",
+            split="train",
+        )
+        packer = TextPacker(padding_idx=0)
+        dataset = PackedDataset(
+            dataset=dataset,
+            packer=packer,
+            target_tokens_per_pack=1024,
```
Comment (felipemello1) [inline apps/sft/sft.py]: do we have it in the cfg?

File: forge/datasets/packed.py (hunk 11 above)
Comment (ebsmothers) [reply forge/datasets/packed.py:640]: For the initial version it was just one less thing to test. Assuming it works smoothly I can add it back in

File: apps/sft/sft.py (hunk 12 above)
Comment (ebsmothers) [reply apps/sft/sft.py]: Discussed this a bit over chat on Friday, I think we will currently need to hardcode this to match model max seq len

File: forge/_cli/download.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (joecummings) [inline forge/_cli/download.py:1]: Let's not bring this over yet?

File: forge/_cli/run.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (joecummings) [inline forge/_cli/run.py:1]: Lets just launch with torchrun for now?

File: README.md [hunk 15]
```diff
@@ -8,13 +8,27 @@
...
-pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/cu126 --force-reinstall
-[For AMD GPU] pip3 install --pre torch --index-url https://download.pytorch.org/whl/nightly/rocm6.3 --force-reinstall
 ```
+
+## Quick Start
+
+To run SFT for Llama3 8B, run
+
+```bash
+forge download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct --ignore-patterns "original/consolidated.00.pth"
```
Comment (joecummings) [inline README.md:28]: huggingface-cli download ...

File: README.md [hunk 16]
```diff
@@ -8,13 +8,27 @@
...
+forge run --nproc_per_node 2 apps/sft/sft.py --config apps/sft/configs/llama3_8b.yaml
```
Comment (joecummings) [inline README.md:29]: torchrun --nproc-per-node apps/sft/main.py --config apps/sft/llama3_8b.yaml

Comment (joecummings) [reply README.md:29]: No more CLI system :/

**Comment by ebsmothers** [comment]
@joecummings are the CLI comments blockers? It runs, right? I think we can separately figure out the CLI experience (e.g. @allenwang28 is already working on some of the Monarch/MAST stuff anyways so I expect there to be some nontrivial changes here regardless)
